src/rust/engine/src/nodes.rs

```diff
@@ -107,165 +107,166 @@ impl StoreFileByDigest<Failure> for Context {
     }
 }
 
+async fn select(
+    context: Context,
+    mut params: Params,
+    entry: Intern<rule_graph::Entry<Rule>>,
+) -> NodeResult<Value> {
+    params.retain(|k| match entry.as_ref() {
```

Comment:
It has, but I don't see a way to avoid that without adding a lot of extraneous code. This code path may be called by a variety of consumers who would want that filtering to happen, whereas the `Root` constructor must do its own pre-filtering to make its own identity small before it is constructed.